BUG: Fix #59429 stacked bar label position

tev-dixon opened the pull request:
- [x] closes #59429
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Commit:
BUG: Fix label positioning in stacked bar plots with zero-height bars (GH#59429)

tev-dixon committed:
Fix pre-commits, add whatsnew documentation

rhshadrach commented on pandas/tests/plotting/frame/test_frame.py:
```diff
@@ -774,6 +774,15 @@ def test_bar_nan_stacked(self):
         assert result == expected
 
+    def test_bar_stacked_label_position_with_zero_height(self):
+        df = DataFrame({"A": [3, 0, 1], "B": [0, 2, 4], "C": [5, 0, 2]})
```
Can you start this test with a comment referencing the GH issue. E.g. see the test immediately below for an example.

rhshadrach commented on doc/source/whatsnew/v2.3.0.rst:
```diff
@@ -143,7 +143,7 @@ Period
 Plotting
 ^^^^^^^^
--
+- Bug in :meth:`DataFrame.plot.bar` with ``stacked=True`` where labels on stacked bars with zero-height segments were incorrectly positioned at the base instead of the label position of the previous segment (:issue:`59429`)
```
This note should be move to v3.0.0.rst.

rhshadrach replied:
2.3.0 is a "special" release in that we are not doing it from the main branch. We are only backporting certain PRs to the 2.3.0 branch.

rhshadrach reviewed (changes requested):
Thanks for the PR!

tev-dixon committed:
Merge branch 'main' of github.com:tev-dixon/pandas into fix-gh59429-stacked-bar-label-position

tev-dixon committed:
Moved whatsnew documentation from v2.3.0 to v3.0.0 as requested, added issue tracking number to test as requested

tev-dixon commented:
Thank you! I made the requested changes.
[x] removed the errant documentation from `doc/source/whatsnew/v2.3.0.rst`
[x] added documentation to `doc/source/whatsnew/v3.0.0.rst`
[x] added the github issue tracking number to the test in `pandas/tests/plotting/frame/test_frame.py`

tev-dixon commented:
I just want to add that my comment on the issue still stands: [comment](https://github.com/pandas-dev/pandas/issues/59429#issuecomment-2458796268)

By making this change, there is a side effect that will effectively bold some of the slice labels where multiple labels now overlap. I'm not sure if this is an intended behavior, and have assumed that it is. More details in my comment.

rhshadrach reviewed (approved):
lgtm

rhshadrach commented:
Thanks @tev-dixon!
